src/uri.rs

```diff
@@ -0,0 +1,87 @@
+use std::{hash::Hash, ops::Deref, str::FromStr};
+
+use serde::{de::Error, Deserialize, Serialize};
+
+/// Newtype struct around `url::Url` with serialization implementations that properly encode brackets for LSP compatibility.
+#[derive(Debug, Clone)]
+pub struct Uri(url::Url);
+
+impl Serialize for Uri {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        // Encode brackets for LSP compatibility
+        let url_str = self.0.as_str();
+        if url_str.contains('[') || url_str.contains(']') {
+            // Fast bracket encoding - count brackets first for precise capacity
+            let bracket_count = url_str.chars().filter(|&c| c == '[' || c == ']').count();
+            let mut encoded = String::with_capacity(url_str.len() + bracket_count * 2);
```

Comment:
```suggestion
            let mut encoded = String::with_capacity(url_str.len() + bracket_count * 3);
```

That has to be `3` given that we push `%5B` and another 3-sized item, right?

I think we can avoid that entirely and use iterators: `url_str.chars().map(|c| ...).collect::<String>()` to do that automatically, without any extra `let bracket_count` calculation and an `count brackets first for precise capacity` odd comment.

Reply:
No it is two because you are replacing one char for two. But the iterator is nicer anyway.